File: src/agent_client.py
import json
import urllib.parse
import urllib.error
import urllib.request

try:
    import websocket
except ImportError as exc:  # pragma: no cover - import error path depends on environment
    raise RuntimeError("Install websocket-client first: pip install websocket-client") from exc


class PredictionMarketsAgentClient:
    def __init__(self, base_url, api_key):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.ws = None

    def auth_headers(self):
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

    def request_json(self, path, method="GET", body=None):
        data = None if body is None else json.dumps(body).encode("utf-8")
        request = urllib.request.Request(
            f"{self.base_url}{path}",
            method=method,
            headers=self.auth_headers(),
            data=data,
        )
        try:
            with urllib.request.urlopen(request) as response:
                return json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as error:
            payload = error.read().decode("utf-8")
            raise RuntimeError(f"{path} failed with {error.code}: {payload}") from error

    def get_status(self):
        return self.request_json("/api/status")

    def get_protocol(self):
        return self.request_json("/api/agent/protocol")

    def get_schemas(self):
        return self.request_json("/api/agent/schemas")

    def get_me(self):
        return self.request_json("/api/me")

    def get_account_summary(self):
        return self.request_json("/api/account/summary")

    def get_markets(self):
        return self.request_json("/api/markets")

    def get_notifications(self):
        return self.request_json("/api/notifications")

    def create_notification(self, payload):
        return self.request_json("/api/notifications", method="POST", body=payload)

    def update_notification(self, notification_id, payload):
        return self.request_json(f"/api/notifications/{urllib.parse.quote(str(notification_id), safe='')}", method="PATCH", body=payload)

    def delete_notification(self, notification_id):
        return self.request_json(f"/api/notifications/{urllib.parse.quote(str(notification_id), safe='')}", method="DELETE")

    def get_current_market(self):
        return self.request_json("/api/markets/current")

    def get_orders(self):
        return self.request_json("/api/orders")

    def get_order(self, order_hash):
        return self.request_json(f"/api/orders/{urllib.parse.quote(str(order_hash), safe='')}")

    def get_order_events(self):
        return self.request_json("/api/order-events")

    def get_fills(self):
        return self.request_json("/api/fills")

    def get_positions(self):
        return self.request_json("/api/positions")

    def get_claims(self):
        return self.request_json("/api/claims")

    def get_nonce_invalidations(self):
        return self.request_json("/api/nonce-invalidations")

    def submit_claim(self, market_id):
        return self.request_json("/api/claims", method="POST", body={"market_id": market_id})

    def place_order_rest(self, payload):
        return self.request_json("/api/orders/place", method="POST", body=payload)

    def cancel_order_rest(self, order_hash):
        return self.request_json("/api/orders/cancel", method="POST", body={"order_hash": order_hash})

    def take_order_rest(self, order_hash, size_shares, taker_address=None):
        payload = {
            "order_hash": order_hash,
            "size_shares": size_shares,
        }
        if taker_address is not None:
            payload["taker_address"] = taker_address
        return self.request_json("/api/orders/take", method="POST", body=payload)

    def build_agent_ws_url(self):
        if self.base_url.startswith("https://"):
            ws_base = "wss://" + self.base_url[len("https://") :]
        else:
            ws_base = "ws://" + self.base_url[len("http://") :]
        return f"{ws_base}/ws/agent?api_key={self.api_key}"

    def connect(self):
        self.ws = websocket.create_connection(self.build_agent_ws_url())
        return self.ws

    def disconnect(self):
        if self.ws is not None:
            try:
                self.ws.close()
            except Exception:
                pass
        self.ws = None

    def send(self, event_type, payload=None):
        if self.ws is None:
            raise RuntimeError("Agent websocket is not connected.")
        self.ws.send(json.dumps({"type": event_type, "payload": payload or {}}))

    def recv(self):
        if self.ws is None:
            raise RuntimeError("Agent websocket is not connected.")
        raw_message = self.ws.recv()
        return json.loads(raw_message)

    def subscribe_book(self, market_id=None, since_sequence=None):
        self.send(
            "book.subscribe",
            {
                "market_id": market_id,
                "since_sequence": since_sequence,
            },
        )

    def subscribe_positions(self, market_id=None, since_sequence=None):
        self.send(
            "positions.subscribe",
            {
                "market_id": market_id,
                "since_sequence": since_sequence,
            },
        )

    def subscribe_nonce_invalidations(self, market_id=None, since_sequence=None):
        self.send(
            "nonce_invalidations.subscribe",
            {
                "market_id": market_id,
                "since_sequence": since_sequence,
            },
        )

    def place_order(
        self,
        market_id,
        side,
        price_tenths,
        size_shares,
        expiry_ms=None,
        client_order_id=None,
        maker_address=None,
        nonce=None,
        signature=None,
    ):
        payload = {
            "market_id": market_id,
            "side": side,
            "price_tenths": price_tenths,
            "size_shares": size_shares,
        }
        if expiry_ms is not None:
            payload["expiry_ms"] = expiry_ms
        if client_order_id is not None:
            payload["client_order_id"] = client_order_id
        if maker_address is not None:
            payload["maker_address"] = maker_address
        if nonce is not None:
            payload["nonce"] = nonce
        if signature is not None:
            payload["signature"] = signature
        self.send("order.place", payload)

    def cancel_order(self, order_hash):
        self.send("order.cancel", {"order_hash": order_hash})

    def take_order(self, order_hash, size_shares, taker_address=None):
        payload = {
            "order_hash": order_hash,
            "size_shares": size_shares,
        }
        if taker_address is not None:
            payload["taker_address"] = taker_address
        self.send("order.take", payload)

File: src/agent-client.ts
export type Json = Record<string, unknown>;

export type AgentScope =
  | "market:read"
  | "orders:write"
  | "orders:cancel"
  | "fills:take"
  | "claims:write"
  | "nonce_invalidations:read"
  | "admin"
  | "*";

export type AgentStreamMode = "snapshot" | "replay" | "snapshot_after_gap";
export type AgentRecoveryReason = "future_sequence" | "gap_out_of_range";

export type AgentStreamSync = {
  stream: string;
  mode: AgentStreamMode;
  from_sequence: number | null;
  to_sequence: number | null;
  server_now_ms: number;
};

export type AgentStreamRecoveryRequired = {
  stream: string;
  requested_sequence: number;
  latest_sequence: number;
  oldest_replayable_sequence: number;
  reason: AgentRecoveryReason;
  server_now_ms: number;
};

export type AgentBootstrapPayload = {
  server_now_ms: number;
  agent_id: string | null;
  market_id: string | null;
  current_round: { endMs?: number } | null;
  streams?: Json;
  auth?: {
    key_id?: string | null;
    key_prefix?: string | null;
    source?: string | null;
    scopes?: AgentScope[];
    rate_limit_per_minute?: number | null;
  };
  capabilities?: Json;
} & Json;

export type AgentWsMessage =
  | { type: "agent.bootstrap"; payload: AgentBootstrapPayload }
  | { type: "stream.sync"; payload: AgentStreamSync }
  | { type: "stream.recovery_required"; payload: AgentStreamRecoveryRequired }
  | { type: "error"; payload: { error?: string } & Json }
  | { type: string; payload?: Json };

export type PredictionMarketsAgentClientOptions = {
  baseUrl: string;
  apiKey: string;
  reconnect?: boolean;
  reconnectDelayMs?: number;
  WebSocketImpl?: typeof WebSocket;
  fetchImpl?: typeof fetch;
};

export type SubscribeStreamOptions = {
  marketId?: string | null;
  sinceSequence?: number | null;
};

export type PlaceOrderPayload = {
  market_id: string;
  side: "buy_yes" | "sell_yes";
  price_tenths: number;
  size_shares: number;
  expiry_ms?: number;
  client_order_id?: string;
  maker_address?: string;
  nonce?: number;
  signature?: string;
};

export class PredictionMarketsAgentClient {
  readonly baseUrl: string;
  readonly apiKey: string;
  readonly reconnect: boolean;
  readonly reconnectDelayMs: number;
  readonly WebSocketImpl: typeof WebSocket;
  readonly fetchImpl: typeof fetch;

  private ws: WebSocket | null = null;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private manuallyClosed = false;

  onOpen: (() => void) | null = null;
  onClose: (() => void) | null = null;
  onError: ((error: unknown) => void) | null = null;
  onMessage: ((message: AgentWsMessage) => void) | null = null;
  onRecoveryRequired: ((payload: AgentStreamRecoveryRequired) => void) | null = null;

  constructor(options: PredictionMarketsAgentClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.apiKey = options.apiKey;
    this.reconnect = options.reconnect ?? true;
    this.reconnectDelayMs = options.reconnectDelayMs ?? 1000;
    this.WebSocketImpl = options.WebSocketImpl ?? WebSocket;
    this.fetchImpl = options.fetchImpl ?? fetch;
  }

  authHeaders(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
    };
  }

  async requestJson<T = Json>(path: string, init: RequestInit = {}): Promise<T> {
    const response = await this.fetchImpl(`${this.baseUrl}${path}`, {
      ...init,
      headers: {
        ...this.authHeaders(),
        ...(init.headers || {}),
      },
    });
    const payload = (await response.json().catch(() => ({}))) as T & { error?: string };
    if (!response.ok) {
      throw new Error(payload.error || `${path} failed with ${response.status}`);
    }
    return payload;
  }

  getStatus() {
    return this.requestJson("/api/status");
  }

  getProtocol() {
    return this.requestJson("/api/agent/protocol");
  }

  getSchemas() {
    return this.requestJson("/api/agent/schemas");
  }

  getMe() {
    return this.requestJson("/api/me");
  }

  getAccountSummary() {
    return this.requestJson("/api/account/summary");
  }

  getMarkets() {
    return this.requestJson("/api/markets");
  }

  getNotifications() {
    return this.requestJson("/api/notifications");
  }

  createNotification(payload: Record<string, unknown>) {
    return this.requestJson("/api/notifications", {
      method: "POST",
      body: JSON.stringify(payload),
    });
  }

  updateNotification(id: string, payload: Record<string, unknown>) {
    return this.requestJson(`/api/notifications/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    });
  }

  deleteNotification(id: string) {
    return this.requestJson(`/api/notifications/${encodeURIComponent(id)}`, {
      method: "DELETE",
    });
  }

  getCurrentMarket() {
    return this.requestJson<{ market_id: string | null }>("/api/markets/current");
  }

  getOrders() {
    return this.requestJson("/api/orders");
  }

  getOrder(orderHash: string) {
    return this.requestJson(`/api/orders/${encodeURIComponent(orderHash)}`);
  }

  getOrderEvents() {
    return this.requestJson("/api/order-events");
  }

  getFills() {
    return this.requestJson("/api/fills");
  }

  getPositions() {
    return this.requestJson("/api/positions");
  }

  getClaims() {
    return this.requestJson("/api/claims");
  }

  getNonceInvalidations() {
    return this.requestJson("/api/nonce-invalidations");
  }

  submitClaim(marketId: string) {
    return this.requestJson("/api/claims", {
      method: "POST",
      body: JSON.stringify({ market_id: marketId }),
    });
  }

  placeOrderRest(payload: PlaceOrderPayload) {
    return this.requestJson("/api/orders/place", {
      method: "POST",
      body: JSON.stringify(payload),
    });
  }

  cancelOrderRest(orderHash: string) {
    return this.requestJson("/api/orders/cancel", {
      method: "POST",
      body: JSON.stringify({ order_hash: orderHash }),
    });
  }

  takeOrderRest(orderHash: string, sizeShares: number, takerAddress?: string) {
    return this.requestJson("/api/orders/take", {
      method: "POST",
      body: JSON.stringify({
        order_hash: orderHash,
        size_shares: sizeShares,
        ...(takerAddress ? { taker_address: takerAddress } : {}),
      }),
    });
  }

  buildAgentWsUrl(): string {
    const wsBase = this.baseUrl.replace(/^http:/, "ws:").replace(/^https:/, "wss:");
    const url = new URL(`${wsBase}/ws/agent`);
    url.searchParams.set("api_key", this.apiKey);
    return url.toString();
  }

  connect() {
    this.manuallyClosed = false;
    this.clearReconnectTimer();
    this.ws = new this.WebSocketImpl(this.buildAgentWsUrl());

    this.ws.addEventListener("open", () => {
      this.onOpen?.();
    });
    this.ws.addEventListener("close", () => {
      this.onClose?.();
      if (this.reconnect && !this.manuallyClosed) {
        this.reconnectTimer = setTimeout(() => this.connect(), this.reconnectDelayMs);
      }
    });
    this.ws.addEventListener("error", event => {
      this.onError?.(event);
    });
    this.ws.addEventListener("message", event => {
      const message = JSON.parse(
        typeof event.data === "string" ? event.data : String(event.data ?? ""),
      ) as AgentWsMessage;
      if (message.type === "stream.recovery_required") {
        this.onRecoveryRequired?.(message.payload);
      }
      this.onMessage?.(message);
    });
  }

  disconnect() {
    this.manuallyClosed = true;
    this.clearReconnectTimer();
    if (this.ws) {
      try {
        this.ws.close();
      } catch {}
    }
    this.ws = null;
  }

  isConnected(): boolean {
    return this.ws?.readyState === this.WebSocketImpl.OPEN;
  }

  send(type: string, payload: Json = {}) {
    if (!this.ws || this.ws.readyState !== this.WebSocketImpl.OPEN) {
      throw new Error("Agent websocket is not connected.");
    }
    this.ws.send(JSON.stringify({ type, payload }));
  }

  subscribeBook(options: SubscribeStreamOptions = {}) {
    this.send("book.subscribe", {
      market_id: options.marketId ?? null,
      since_sequence: options.sinceSequence ?? null,
    });
  }

  subscribePositions(options: SubscribeStreamOptions = {}) {
    this.send("positions.subscribe", {
      market_id: options.marketId ?? null,
      since_sequence: options.sinceSequence ?? null,
    });
  }

  subscribeNonceInvalidations(options: SubscribeStreamOptions = {}) {
    this.send("nonce_invalidations.subscribe", {
      market_id: options.marketId ?? null,
      since_sequence: options.sinceSequence ?? null,
    });
  }

  placeOrder(payload: PlaceOrderPayload) {
    this.send("order.place", payload as unknown as Json);
  }

  cancelOrder(orderHash: string) {
    this.send("order.cancel", { order_hash: orderHash });
  }

  takeOrder(orderHash: string, sizeShares: number, takerAddress?: string) {
    this.send("order.take", {
      order_hash: orderHash,
      size_shares: sizeShares,
      ...(takerAddress ? { taker_address: takerAddress } : {}),
    });
  }

  private clearReconnectTimer() {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }
}
